### DISCUSSION TITLE
Notifications are not instant?

### DISCUSSION THREAD

**Discussion by tonaltti** [SeoulSKY/ytnoti#1] [General]
If I understand correctly, PubSubHubbub is based on YouTube's RSS feed. In my experience, the RSS feed updates anywhere from 0 to 20 minutes after an upload, so the notification is far from instant.

Is there a way to get instant notifications about uploads?

**Comment by SeoulSKY** [comment]
I couldn't find a reference that YouTube's PubSubHubbub is implemented with an RSS feed on their server.

The difference between the RSS feed and PubSubHubbub is that the RSS feed requires a client to do periodic polling, whereas, for PubSubHubbub, a client gets a POST request from the server. So, in theory, the notification from PubSubHubbub should be instant, unless the YouTube server uses a mechanism that introduces some delay.

I haven't measured the delay for their PubSubHubbub, but if you want to control the delay, you can use their RSS feed with periodic polling with a short interval, although it will be less efficient than using PubSubHubbub

**Reply by tonaltti** [reply]
With RSS feed I mean the video feed:

https://github.com/SeoulSKY/ytnoti/blob/af814dc5d1ca7c6c7f527a71e08da0dedd62e13f/ytnoti/__init__.py#L484

If I poll this address every minute: https://www.youtube.com/feeds/videos.xml?channel_id=UC9EEyg7QBL-stRX-7hTV3ng

Then the new uploads appears to the feed with 0 to 20 minute delay.

It might be possible that PubSubHubbub sends a notification about new upload even when the video is not in the feed yet. I was going to test that, but when I run following script:

``` py
from pyngrok import ngrok
from ytnoti import YouTubeNotifier, Video
from datetime import datetime

ngrok.set_auth_token("Your ngrok token here")

notifier = YouTubeNotifier()

@notifier.upload()
async def listener(video: Video):
    timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    print(f"[{timestamp}] New video from {video.channel.name}: {video.title}")

notifier.subscribe("UC9EEyg7QBL-stRX-7hTV3ng")
notifier.run()
```

I get following output:

```
Task exception was never retrieved
future: <Task finished name='Task-4' coro=<BaseYouTubeNotifier._get_server.<locals>.on_ready() done, defined at C:\Python\Lib\site-packages\ytnoti\base.py:320> exception=ValueError('Invalid channel ID: UC9EEyg7QBL-stRX-7hTV3ng')>
Traceback (most recent call last):
  File "C:\Python\Lib\site-packages\ytnoti\base.py", line 326, in on_ready
    await self._register(self._subscribed_ids)
  File "C:\Python\Lib\site-packages\ytnoti\base.py", line 396, in _register
    raise ValueError(f"Invalid channel ID: {channel_id}")
ValueError: Invalid channel ID: UC9EEyg7QBL-stRX-7hTV3ng
```

The channel exists. So the problem has to be somewhere else. It might be related to my firewall config (I understood this approach requires that my machine must be reachable from the Internet), but even then the error message is incorrect. I also just clean installed Python 3.12.5 but the problem persists.

**Reply by SeoulSKY** [reply]
When it verifies a CHANNEL_ID, it sends a HEAD request to https://www.youtube.com/channel/CHANNEL_ID and checks if the response code is 200.

Maybe the request returned something other than 200 for some reason (Maybe bot protection?). In the next update, it will use the feed URL to verify the channel ID. Feed URL shouldn't block bots

**Reply by tonaltti** [reply]
At least Postman returns 200 OK from https://www.youtube.com/channel/UC9EEyg7QBL-stRX-7hTV3ng HEAD request.

**Reply by SeoulSKY** [reply]
I tried it with Insomnia (which is an alternative to Postman) and it works too.

Assuming you're running your Python code in a cloud computing instance, you can try running these on your server and see if they work:
```bash
curl -I "https://www.youtube.com/channel/UC9EEyg7QBL-stRX-7hTV3ng"
curl -I "https://www.youtube.com/feeds/videos.xml?channel_id=UC9EEyg7QBL-stRX-7hTV3ng"
```

The first link is the one that the library is using currently, and the second one is for the next update.

**Reply by tonaltti** [reply]
Both returns 200 OK.

**Reply by SeoulSKY** [reply]
I'm going to publish a new update shortly and let's see if the issue persists after that

**Reply by SeoulSKY** [reply]
Check out the new update:

```bash
pip install -U ytnoti
```

**Reply by tonaltti** [reply]
Now I do not get any output at all. Maybe it's listening for new videos. I need to now catch an upload from some channel.

**Reply by SeoulSKY** [reply]
Great!

**Reply by tonaltti** [reply]
It works, I get notifications about new uploads.

And also the notifications are instant. The delay to published_at timestamp from Youtube API is only 3-4 seconds.

Interestingly the video is still added to the feed file https://www.youtube.com/feeds/videos.xml?channel_id=UC9EEyg7QBL-stRX-7hTV3ng with a delay. In my latest test the delay was over 5 minutes.

But this is not a problem to me. I get most of the details in the notification video object and the missing details (like getting the type of the video: normal, live (upcoming), live (currently live), live (completed)) I can then query from Youtube API.

Maybe a support (optional) to Youtube API to get all details about the video could be integrated to this library?

I'll be using this library and I will report if I find any other problems or if I have any further suggestions. Thanks.

**Reply by SeoulSKY** [reply]
I'm glad that you successfully received the notification!

I can add an option to fetch additional video information from YouTube Data API in the next version.

Thank you for the suggestion and feel free to open up issues whenever you encounter a problem!

**Reply by GodlyV** [reply]
> It works, I get notifications about new uploads.
> 
> And also the notifications are instant. The delay to published_at timestamp from Youtube API is only 3-4 seconds.
> 
> Interestingly the video is still added to the feed file https://www.youtube.com/feeds/videos.xml?channel_id=UC9EEyg7QBL-stRX-7hTV3ng with a delay. In my latest test the delay was over 5 minutes.
> 
> But this is not a problem to me. I get most of the details in the notification video object and the missing details (like getting the type of the video: normal, live (upcoming), live (currently live), live (completed)) I can then query from Youtube API.
> 
> Maybe a support (optional) to Youtube API to get all details about the video could be integrated to this library?
> 
> I'll be using this library and I will report if I find any other problems or if I have any further suggestions. Thanks.

Hi, 

Can you clarify what you mean by And also the notifications are instant. The delay to published_at timestamp from Youtube API is only 3-4 seconds? please

**Comment by tonaltti** [comment]
I have my script running since yesterday. Yesterday it notified about the uploads. Twice about the same live stream, but at least I got the notification.

But today it has not been giving notifications. No errors, nothing. What is the best way to debug this? It's important that the library would be reliable and it should at least throw an exception if something is not right. This is my current script:

```py
from pyngrok import ngrok
from ytnoti import YouTubeNotifier, Video
from datetime import datetime

channel_ids = [
    "channelid1",
    "channelid2",
    "channelid3",
    "channelid4",
    "channelid5"
]

ngrok.set_auth_token("mytoken")

notifier = YouTubeNotifier()

@notifier.upload()
async def listener(video: Video):
    timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    print(f"[{timestamp}] New video properties:")
    for key, value in vars(video).items():
        print(f"{key}: {value}")

notifier.subscribe(channel_ids)

notifier.run()
```

**Reply by SeoulSKY** [reply]
It should throw an exception if something is not right. Maybe you simply didn't get any notifications.

If you don't setup the logger, it doesn't print anything by default.

You can set up the logging level to debug level.

```python
import logging

logging.basicConfig(level=logging.DEBUG)
```

**Reply by tonaltti** [reply]
I got the problem again. I left my script with debug logger running for 9 days. On the 6th day I got following log to the console:

```
2024-09-10 22:48:29,211 - httpx - DEBUG - load_ssl_context verify=True cert=None trust_env=True http2=False
2024-09-10 22:48:29,224 - httpx - DEBUG - load_verify_locations cafile='C:\\Python\\Lib\\site-packages\\certifi\\cacert.pem'
2024-09-10 22:48:29,451 - YouTubeNotifier - DEBUG - Sending subscribe request for channel: MY_CHANNEL_ID
2024-09-10 22:48:29,455 - httpcore.connection - DEBUG - connect_tcp.started host='pubsubhubbub.appspot.com' port=443 local_address=None timeout=5.0 socket_options=None
2024-09-10 22:48:29,504 - httpcore.connection - DEBUG - connect_tcp.complete return_value=<httpcore._backends.anyio.AnyIOStream object at 0x000001A381BEE5D0>
2024-09-10 22:48:29,504 - httpcore.connection - DEBUG - start_tls.started ssl_context=<ssl.SSLContext object at 0x000001A3817B1A50> server_hostname='pubsubhubbub.appspot.com' timeout=5.0
2024-09-10 22:48:29,533 - httpcore.connection - DEBUG - start_tls.complete return_value=<httpcore._backends.anyio.AnyIOStream object at 0x000001A381BFD2B0>
2024-09-10 22:48:29,534 - httpcore.http11 - DEBUG - send_request_headers.started request=<Request [b'POST']>
2024-09-10 22:48:29,534 - httpcore.http11 - DEBUG - send_request_headers.complete
2024-09-10 22:48:29,535 - httpcore.http11 - DEBUG - send_request_body.started request=<Request [b'POST']>
2024-09-10 22:48:29,535 - httpcore.http11 - DEBUG - send_request_body.complete
2024-09-10 22:48:29,535 - httpcore.http11 - DEBUG - receive_response_headers.started request=<Request [b'POST']>
2024-09-10 22:48:30,526 - httpcore.http11 - DEBUG - receive_response_headers.complete return_value=(b'HTTP/1.1', 409, b'Conflict', [(b'Cache-Control', b'no-cache'), (b'Content-Type', b'text/plain; charset=utf-8'), (b'X-Cloud-Trace-Context', b'CENSORED'), (b'Date', b'Tue, 10 Sep 2024 19:48:30 GMT'), (b'Server', b'Google Frontend'), (b'Content-Length', b'21'), (b'Alt-Svc', b'h3=":443"; ma=2592000,h3-29=":443"; ma=2592000')])
2024-09-10 22:48:30,531 - httpx - INFO - HTTP Request: POST https://pubsubhubbub.appspot.com "HTTP/1.1 409 Conflict"
2024-09-10 22:48:30,532 - httpcore.http11 - DEBUG - receive_response_body.started request=<Request [b'POST']>
2024-09-10 22:48:30,534 - httpcore.http11 - DEBUG - receive_response_body.complete
2024-09-10 22:48:30,535 - httpcore.http11 - DEBUG - response_closed.started
2024-09-10 22:48:30,536 - httpcore.http11 - DEBUG - response_closed.complete
2024-09-10 22:48:30,537 - httpcore.connection - DEBUG - close.started
2024-09-10 22:48:30,538 - httpcore.connection - DEBUG - close.complete
2024-09-10 22:48:30,544 - asyncio - ERROR - Task exception was never retrieved
future: <Task finished name='Task-8' coro=<AsyncYouTubeNotifier._repeat_task() done, defined at C:\Python\Lib\site-packages\ytnoti\__init__.py:435> exception=HTTPError('Failed to subscribe channel: MY_CHANNEL_ID. The reason might be because https://CENSORED.ngrok-free.app is inaccessible from the public internet', 409)>
Traceback (most recent call last):
  File "C:\Python\Lib\site-packages\ytnoti\__init__.py", line 448, in _repeat_task
    await task()
  File "C:\Python\Lib\site-packages\ytnoti\__init__.py", line 421, in task
    await self._register(self._subscribed_ids)
  File "C:\Python\Lib\site-packages\ytnoti\__init__.py", line 664, in _register
    raise HTTPError(
ytnoti.errors.HTTPError: Status code: 409: Failed to subscribe channel: MY_CHANNEL_ID. The reason might be because https://CENSORED.ngrok-free.app is inaccessible from the public internet
```

After this error the script didn't output anything and it didn't notify about any new videos.

**Reply by SeoulSKY** [reply]
I assume that the problem occurred because your ngrok tunnel was disconnected for some reason (like losing internet connection) and the url is no longer valid.

The url that ngrok generates is temporary and becomes invalid after the tunnel disconnects.

I'd highly recommend that you use ngrok for development only and use your server's domain for production.

Still, I'll investigate more about this.

**Reply by tonaltti** [reply]
That makes sense. Ngrok provides a free static domain so I got one. But how do I use it with ytnoti?

I tried like this:

``` py
tunnel = ngrok.connect(
    addr=8000,
    proto="http",
    domain="mystaticdomain.ngrok-free.app"
)

notifier = YouTubeNotifier()
```
And it opens a tunnel with my domain. But later I see in the log that it opens a tunnel again without the static domain. So I guess ytnoti needs somekind of support for this?

**Reply by tonaltti** [reply]
Nevermind. I can do this:

```py
tunnel = ngrok.connect(
    addr=8000,
    proto="http",
    domain="mystaticdomain.ngrok-free.app"
)

notifier = YouTubeNotifier(callback_url="https://mystaticdomain.ngrok-free.app")
```

Now lets see if it works for a week.

**Reply by SeoulSKY** [reply]
Great! Hopefully it works

**Reply by tonaltti** [reply]
Just an update that with static domain I didn't face the same problem anymore. But maybe the dynamic ngrok domain problem could be described in the documentation as that is something many users are likely to experience.

**Reply by SeoulSKY** [reply]
Yeah, I'll update the document so that they are aware it's for development only
